src/index.js

```diff
@@ -883,34 +929,49 @@ export default function (babel) {
       },
 
       ArrayComprehension(path) {
-        // disallow Yield and Return
-        path.get("loop.body").traverse({
-          AwaitExpression(awaitPath) {
-            throw awaitPath.buildCodeFrameError(
-              "`await` is not allowed within Comprehensions; " +
-              "instead, await the Comprehension (eg; `y <- [for x of xs: x]`)."
-            );
-          },
-          YieldExpression(yieldPath) {
-            throw yieldPath.buildCodeFrameError("`yield` is not allowed within Comprehensions.");
-          },
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
```

Comment:
@rattrayalex 

Yup, I think this can't be an exit visitor. (At least not without further extensive modifications.) The `ForInArrayStaetement` gets transformed first and that breaks the assumptions we're making about ast pretty badly.

I have a working implemenation as an enter visitor which I think I will stick with for this PR.